### ISSUE TITLE
Bind actions instead of coupling them to the store

### ISSUE THREAD

**Issue by developit** [concretesolutions/redux-zero#16]
Hiya!

Right now, actions must import an instance of the store in order to invoke `setState()`. This creates a tight coupling of actions to the store, making testing require mocking that import.  It's a short-circuit that starts out quite nice (just import the store instance), but can get a bit hairy later on (can only have one store, can't compose actions, etc).

Here's a testing use-case that shows what I mean:

```js
import { increment, decrement } from './actions'
import store from './store'  // not part of the unit being tested

increment()
expect(store.getState()).to.eql({ count: 1 })   // asserting on another module
// ^ what if something else had changed the store? or messed with store.setState?
```

I'd like to propose a few lightweight options for making actions "late-bound".  Some of this comes from other experiments I've done with the Gist I sent.

> _**Note:** in the examples I'm using `Object` in place of `mapStateToProps` - I'm just using it as the identity function so that the entire store state is passed as props for the examples._

# 1. Simple store binding function

This is fairly analogous to redux's approach, just replacing `dispatch()` with the store instance itself.  `connect()` accepts a function as a 2nd argument that expects `(store)` and returns actions that mutate that store.  Perks: this remains compatible and nearly identical to the readme examples, the only difference being that `store` is injected rather than imported (therein removing the singleton issue).

```js
import { createStore, Provider, connect } from 'redux-zero'
export default () => <Provider store={createStore()}><App /></Provider>

const mapStateToProps = Object  // or whatever

// same as actions.js from the readme, just instantiable with any `store`:
const createActions = store => ({
  increment() {
    // arguments passed to props.increment() are forwarded here
    store.setState({
      count: store.getState().count + 1
    })
  }
})

const App = connect(mapStateToProps, createActions)(
  ({ count, increment }) => (
    <button onClick={increment}>{count}</button>
  )
)
```

#### Implementation

This one is simple: add a second function argument to `connect()`, then we just call it when constructing the class and pass it the store. That method returns us an object we can pass down as handler props.

```js
export default function connect(mapToProps, createActions) {
  return Child =>
    class Connected extends React.Component {
      // pass the store to createActions() and get back our bound actions:
      actions = createActions(this.context.store)
      // ...
      render() {
        return <Child store={this.context.store} {...this.actions} {...this.props} {...this.state} />
      }
```

#### Testing Example

Since our straw-man was testing, here's how you'd test with the above changes:

```js
import createActions from './actions'
import createStore from 'redux-zero'

// we're importing lots still, but it's all (repeatedly) instantiable.
// Because we instantiated it, we know it's clean (no need to "reset").
let store = createStore({ count: 0 })
let actions = createActions(store)
actions.increment()
expect(store.getState()).to.eql({ count: 1 })
// ^ we know nothing outside could have affected this, it's our test store.
```

# 2. Auto-binding an actions map

This one is nice - instead of actions mutating the store directly, they get "bound" to the store by `connect()`. Now all they have to do is accept `(state, ...args)` as arguments and return a new state - everything else is automated.

It's interesting to note - redux exposes a `bindActionCreators()` method that does this, but it also includes this binding behavior by default when passing an object 2nd argument to `connect()`.

```js
import { createStore, Provider, connect } from 'redux-zero'
export default () => <Provider store={createStore()}><App /></Provider>

const mapStateToProps = Object  // or whatever

// Actions are just pure functions that return a derivative state.
// Importantly, they don't need "hardwired" access to the store.
// Any arguments passed to the bound action are passed after state
const increment = ({ count }) => ({ count: count+1 })

const actions = { increment }

const App = connect(mapStateToProps, actions)(
  ({ count, increment }) => (
    <button onClick={increment}>{count}</button>
  )
)
```

#### Implementation

This adds one small step to `connect()` - in [Connected](https://github.com/concretesolutions/redux-zero/blob/master/src/components/connect.tsx#L7), we'd need to loop over the passed actions and create their store-bound proxies, then store the proxy function mapping as a property of the class.  This could either happen in the constructor, as an instance property, or within componentWillMount.

```js
function bindActions(actions, store) {
  let bound = {}
  for (let name in actions) {
    bound[name] = (...args) => {
      store.setState( actions[name](store.getState(), ...args) )
    }
  }
  return bound
}

export default function connect(mapToProps, actions) {
  return Child =>
    class Connected extends React.Component {
      // bind the pure actions to our store:
      actions = bindActions(actions, this.context.store)
      // ...
      render() {
        return <Child store={this.context.store} {...this.actions} {...this.props} {...this.state} />
      }
```

#### Testing Example

The benefits of this approach get pretty clear when you look at how to test things.  Since actions are now just pure functions that accept the current state and return a new one, they're easy to test in complete isolation:

```js
import { increment, decrement } from './actions'  // not tied to a component

// no store instance needed
expect( increment({ count:0 }) ).to.eql({ count: 1 });
// ^ pure tests are almost pointlessly easy
```

One caveat with this approach is that async actions become a little harder.  With option #1, actions can invoke `store.setState()` in response to async stuff, it'll just always be available.  With option #2, there'd need to be some way to access the store later on.  Another alternative would be to change `bindActions()` to handle Promise return values from actions.

---

Hope this is useful, I'd love to see an option like this make its way into redux-zero!

**Comment by malbernaz** [comment]
@developit your second approach looks awesome! Love the idea of having actions as pure functions. Thing is that you are right about the caveat you've pointed out, we'd have to find a way to handle async actions properly.

I like the option of having bindActions to handle promises though, we just have to figure out how (don't thing it would be too hard).

**Comment by matheusml** [comment]
Although #1 is really simple, I don't think it's going to solve our problem here: we're still having the store somehow coupled with the actions, right?

#2 is just beatiful. Very concise API, just easy to test/reuse pure functions. But the async part is bothering me. Is there a simple way for us to handle it?

**Comment by malbernaz** [comment]
```js
function bindActions(actions, store) {
  let bound = {};

  for (let name in actions) {
    bound[name] = (...args) => {
      const result = actions[name](store.getState(), ...args);

      if (result instanceof Promise) {
        return result.then(store.setState);
      }

      return Promise.resolve().then(() => {
        store.setState(result);
      });
    };
  }

  return bound;
}
```

@matheusml I guess that would be OK. Can someone see any issues?

**Comment by developit** [comment]
@malbernaz I'd avoid waiting on a Promise in the case where no thennable is returned:

```js
function bindActions(actions, store) {
  let bound = {};

  for (let name in actions) {
    bound[name] = (...args) => {
      const result = actions[name](store.getState(), ...args);

      if (result && result.then) {
        result.then(store.setState);
      }
      else {
        store.setState(result);
      }
    };
  }

  return bound;
}
```

**Comment by malbernaz** [comment]
Fair enough... It probably makes more sense

**Comment by developit** [comment]
@matheusml #1 goes further toward solving this problem than it looks at first blush:  it still binds actions to a store, but it does it in a way that can be instantiated.  Here's the proof:

```js
const createActions = store => ({
  increment() {
    store.setState({ count: store.getState().count+1 })
  }
})

// a truly isolated test:
let state = {}
let store = { getState: ()=>state, setState: o=>state={...state,...update} }
let actions = createActions(store)
actions.increment()
expect(state.count).to.equal(1)
```

There are slight variants on this option that highlight the compelling nature a bit better:

```js
// is it a Store instance, or just any two methods? 💁
const createActions = ({ getState, setState }) => ({
  increment() {
    setState({ count: getState().count+1 })
  }
})
```

One last huge advantage here: state updates in this option are _atomic_.  It's very unlikely you'd end up with 2 async actions overwriting eachother while executing.  Contrast that to option #2, which passes the current state to a function that could take 30 seconds to return, and then applies the result of that function to the store - the store could have been mutated during that time, but because the state got passed in at invocation time, those changes won't be reflected and can be overwritten.

Also, it's probably worth calling out that the two options are not mutually exclusive - a type check on the second argument to `connect()` allows for both.  Really it comes down to which is the more powerful primitive.

Personally, I think that the nature of this library is really geared towards `getState()` -> `setState()`.  It's super easy to write update helpers around that:

```js
const pureActions = actions => store => {
  let pure = {}
  for (let i in actions) pure[i] = (...args) => {
    let r = actions[i](store.getState(), ...args)
    if (r && r.then) r.then(store.setState)
    else store.setState(r);
  }
  return pure;
}

const actions = {
  // pure action:
  increment: ({ count }) => ({ count: count+1 }),
  // an async action (note: has same locking issue):
  fetchNewCount: () => fetch('/count').then(r=>r.json()).then( count => ({ count }) )
}

connect(Object, pureActions(actions))( ... )
```

**Comment by matheusml** [comment]
Thanks for the clarification @developit. That was awesome!

That's a really thought choice but I'm more towards to #2. Pure functions seem to me the way to go. Easier to test and to reason about.

With that said, this is a little concerning:

> One last huge advantage here: state updates in this option are atomic. It's very unlikely you'd end up with 2 async actions overwriting eachother while executing. Contrast that to option #2, which passes the current state to a function that could take 30 seconds to return, and then applies the result of that function to the store - the store could have been mutated during that time, but because the state got passed in at invocation time, those changes won't be reflected and can be overwritten.

Correct if I'm wrong, but if we can't guarantee that 2 async functions can properly update the state, this seems like a deal breaker.

**Comment by developit** [comment]
Yup - I would say #2 is ideal for synchronous updates, and #1 is ideal for asynchronous updates.  

Here's a hybrid of both approaches:

```js
function bindActions(actions, store) {
  let bound = {}
  for (let name in actions) {
    bound[name] = (...args) => {
      let ret = actions[name](store.getState(), ...args)
      if (ret!=null) store.setState(ret)
    }
  }
  return bound
}

export default function connect(mapToProps, actions) {
  return Child =>
    class Connected extends React.Component {
      constructor(_, { store }) {
        super()
        this.actions = bindActions(typeof actions==='function' ? actions(store) : actions, store)
      }
      // ...
      render() {
        return <Child store={this.context.store} {...this.actions} {...this.props} {...this.state} />
      }
```

It works with both approaches mixed:

```js
// can be an object w/ pure functions,
// *or* an actions factory that accepts the store
const createActions = store => ({
  // example of a pure action
  increment: state => ({ count: state.count + 1 }),

  // state is always the first argument
  increment(state) {
    store.setState({ count: state.count + 1 })
    // no return value = no implicit state update
  },

  // example async action
  increment(state) {
    setTimeout( () => {
      // can use `state` here, or re-acquire state (better) via getState():
      store.setState({ pending: false, count: store.getState().count + 1 })
    })
    // can combine: here we return the intermediary state update:
    return { pending: true }
  }
})

const App = connect(Object, createActions)(
  ({ count, increment }) => (
    <button onClick={increment}>{count}</button>
  )
)
```

Personally I think this mixed approach ends up being ideal: it avoids the easy-but-broken implicit async via promises approach, allows for simple pure synchronous actions, and simplifies async.  Best of all, every feature is entirely opt-in!

**Comment by malbernaz** [comment]
I'm now leaning more towards the first approach because it's easier to reason about, and because it has that async friendliness that #2 doesn't.

But It's probably best to have a bitter complex API to get the best of both worlds. I think we shoudl implement that.

We'll have to work harder on docs : )

**Comment by matheusml** [comment]
I really liked this #3 approach by @developit

**Comment by matheusml** [comment]
Let me introduce you to alternative #4 

So, I stumbled to a library called [elfi](https://github.com/madx/elfi) which is incredibly similar to Redux Zero. Elfi's approach is simple: store dispatch actions. And actions are functions that have as an input the current state, and returns the modified state (pure functions \o/).

This is how it works:

```javascript
import {createStore} from "elfi"

const store = createStore(1)

function increment(state) {
  return state + 1
}

store.dispatch(increment)
store.getState() // => 2
```
 You can read more about it [here](http://madx.me/articles/a-simpler-alternative-to-flux-and-redux.html)

What you guys @developit, @malbernaz think?

**Comment by eirslett** [comment]
Maybe actions can also be generators instead of just pure functions, then they can do side effects as well:

```javascript
const store = createStore({ loading: false, loaded: false, error: null, data: null });

function* getData() {
  yield state => { ...state, loading: true }
  try {
    const response = yield fetch('bla bla bla')
    const json = yield response.json()
    yield state => { ...state, loading: false, loaded: true, error: null, data: json }
  } catch (err) {
    yield state => { ...state, loading: false, loaded: false, error: err }
  }
}

store.dispatch(getData())
```

Pure generator functions are also easy to unit test, just like redux sagas.

**Comment by developit** [comment]
@matheusml Definitely an interesting option having them be associated with the store. I guess that decision comes down to how you view `redux-zero` being used. For me, the Gist I had originally sent was just a way to bind to shared state, and update it ad-hoc.  Adding actions bound via connect() helps standardize them and create more self-documenting code, and it allows actions to be shared as modules (objects with pure functions that can be passed to connect).

Moving actions to be centralized with the Store starts to move more in the direction of redux itself - it starts to look more like a centralized "brain" for your app, where the components are just dumb views. I've worked on apps where that model is used (generally with Redux) and it's nice, but I'd wonder- is that now too much of a "solved" problem?  Just some pondering! :)

**Comment by matheusml** [comment]
Hey guys, just opened the PR. 
Option #3 it is. Why?
- Pure functions
- Sync and async enabled
- Easy API

Could you ( @malbernaz, @developit, @eirslett )  please take a look?
